stan/math/prim/mat/fun/cholesky_decompose_gpu.hpp

```diff
@@ -20,117 +22,179 @@
+     * @param block size of the block for each step
+     * @return Square root of matrix on the GPU.
+     * @throw std::domain_error if m is not
+     *  positive definite (if m has more than 0 elements)
+     */    
+    matrix_gpu cholesky_decompose_gpu(matrix_gpu & A, int block) {
+      if (A.size() == 0) return A;
+      cl::Kernel kernel_chol_block = get_kernel("cholesky_block");
+      cl::CommandQueue cmd_queue = get_queue();
+      try {
+        // Will be managed by the library core system
+        int offset = 0;
+        matrix_gpu V(block, block);
+        matrix_gpu D(block, block);
+        matrix_gpu Mid;
+        while ((offset + block) < (A.rows())) {
+          matrix_gpu L(A.rows()-offset-block, block);
+          matrix_gpu Mid_temp(A.rows()-offset-block, A.rows()-offset-block);
+
+          copy_submatrix(A, D, offset, offset, 0, 0, block, block);
+          zeros(V);
+          kernel_chol_block.setArg(0, V.buffer());
+          kernel_chol_block.setArg(1, D.buffer());
+          kernel_chol_block.setArg(2, block);
+          cmd_queue.enqueueNDRangeKernel(kernel_chol_block,
+           cl::NullRange, cl::NDRange(block), cl::NDRange(block));
+          copy(V, D);
+          copy_submatrix(V, A, 0, 0, offset, offset, block, block);
+
+          V = lower_triangular_inverse(D);
 
+          copy_submatrix(A, L, (offset+block), offset, 0, 0,
+            (A.rows()-offset-block) , block);
+          V = transpose(V);
+          L = multiply(L, V);
+          copy_submatrix(L, A, 0, 0, (offset+block), offset,
+            (A.rows()-offset-block) , block);
+
+          copy_submatrix(A, Mid_temp, (offset+block), (offset+block),
+            0, 0, (A.rows()-offset-block), (A.rows()-offset-block));
+          Mid = multiply_with_self_transposed(L);
+          Mid = subtract(Mid_temp, Mid);
+          copy_submatrix(Mid, A, 0, 0, (offset+block), (offset+block),
+            (A.rows()-offset-block), (A.rows()-offset-block));
+          offset += block;
+        }
+        int left = A.rows() - offset;
+        if (left > 0) {
+          matrix_gpu D(left, left);
+          matrix_gpu V(left, left);
+          copy_submatrix(A, D, offset, offset, 0, 0, left, left);
+          zeros(V);
+          kernel_chol_block.setArg(0, V.buffer());
+          kernel_chol_block.setArg(1, D.buffer());
+          kernel_chol_block.setArg(2, left);
+          cmd_queue.enqueueNDRangeKernel(kernel_chol_block,
+           cl::NullRange, cl::NDRange(left), cl::NDRange(left));
+           copy_submatrix(V, A, 0, 0, offset, offset, left, left);
+        }
+        zeros(A, UPPER);
+      } catch (const cl::Error& e) {
+        check_ocl_error("cholesky_decompose", e);
+      }
+      copy_triangular_transposed(A, LOWER_TO_UPPER_TRIANGULAR);
+      check_positive_definite_gpu("cholesky_decompose_gpu",
+        "Matrix m", A);
+      zeros(A, UPPER);
+      matrix_gpu B(A.rows(), A.cols());
+      copy(A, B);
```

Comment:
This might be clearer as a copy constructor

```
matrix_gpu B(A);
```

so you'd define

```
matrix _gpu(const matrix_gpu& A);
```
if you need something other than default behavior.